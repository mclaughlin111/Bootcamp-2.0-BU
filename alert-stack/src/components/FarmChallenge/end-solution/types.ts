// Union literals because types should always be constrained as much as possible
// https://www.codecademy.com/learn/learn-typescript/modules/learn-typescript-union-types/cheatsheet
// If someting can only return 5 values, its type should not be "string"
export type AnimalType = "chicken" | "cow" | "goat" | "pig" | "sheep";
export type SortByOptions = "age" | "name" | "worth";

// This is a "helper type" for when we have a union literal with a display name (animal type + sort By Opyions)
export type OptionInfo<T> = {
  type: T;
  displayName: string;
};

// Modelling an animal (and getting more use out of out AnimalType)
export type Animal = {
  _id: string;
  isAlive: boolean;
  worth: number;
  age: number;
  name: string;
  gender: "male" | "female";
  animalType: AnimalType;
};
