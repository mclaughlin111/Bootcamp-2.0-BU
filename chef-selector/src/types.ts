const jobTitles = [
  "Head Chef",
  "Sous Chef",
  "Executive Chef",
  "Station Chef",
  "Pastry Chef",
  "Sauce Chef",
] as const;
export type ChefJobTitle = (typeof jobTitles)[number];

export type MichelinStars = 0 | 1 | 2 | 3;

export type Chef = {
  id: number;
  name: string;
  jobTitle: ChefJobTitle;
  michelinStars: MichelinStars;
  trainingProgress: number;
};

export type ChefResult = Pick<Chef, "id" | "name" | "jobTitle"> & {
  isSuccess: boolean;
};
