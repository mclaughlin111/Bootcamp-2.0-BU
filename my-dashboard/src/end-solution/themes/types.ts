export type ThemeOption = "light" | "dark" | "birindar" | "ugly" | "bad";

export type ThemeConfig = {
  primary: string;
  background: string;
  nav: string;
  border: string;
  text: string;
  sidebar: string;
};
