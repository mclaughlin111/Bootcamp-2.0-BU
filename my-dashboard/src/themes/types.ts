export type ThemeOption = "light" | "dark" | "horriblyunreadable";
