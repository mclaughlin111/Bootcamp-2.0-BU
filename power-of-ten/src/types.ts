export type TimeRow = {
  rank: number;
  chipTime: number;
  gunTime?: number;
  pbTime?: number;
  isPB: boolean;
  name: string;
  grading?: string;
  coach?: string;
  club?: string;
  venue: string;
  date: number;
};
