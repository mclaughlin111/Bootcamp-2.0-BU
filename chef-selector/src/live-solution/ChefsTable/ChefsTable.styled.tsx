import { LinearProgress } from "@mui/material";
import styled from "styled-components";

export const StyledLinearProgress = styled(LinearProgress)`
  width: 100px;
`;
