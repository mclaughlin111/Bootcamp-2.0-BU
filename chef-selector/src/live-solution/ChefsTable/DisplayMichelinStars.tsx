import { MichelinStars } from "../../types";

export const DisplayMichelinStars = ({ stars }: { stars: MichelinStars }) => {
  return (
    <>
      {Array.from(new Array(stars), (_, index) => (
        <img
          key={index}
          src="/michelin-star.png"
          width={24}
          alt="michelin star"
        />
      ))}
    </>
  );
};
