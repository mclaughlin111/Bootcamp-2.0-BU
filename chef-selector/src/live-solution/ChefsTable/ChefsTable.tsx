import {
  Box,
  Checkbox,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from "@mui/material";
import { StyledButton } from "../Button/StyledButton.styled";
import { StyledTableContainer } from "../Table/StyledTableContainer.styled";
import { Chef } from "../../types";
import { useState } from "react";
import { DisplayMichelinStars } from "./DisplayMichelinStars";

type ChefsTableProps = {
  buttonText: string;
  chefs: Chef[];
  onSelected: (selectedIds: number[]) => void;
};

export const ChefsTable = ({
  buttonText,
  chefs,
  onSelected,
}: ChefsTableProps) => {
  const [selectedIds, setSelectedIds] = useState<number[]>([]);

  const onBoxChecked = (chefId: number, isChecked: boolean) => {
    if (isChecked) {
      setSelectedIds((previous) => [...previous, chefId]);
    } else {
      setSelectedIds((previous) => previous.filter((id) => id !== chefId));
    }
  };

  return (
    <>
      <StyledTableContainer>
        <Table sx={{ minWidth: 650 }}>
          <TableHead>
            <TableRow>
              <TableCell>Name</TableCell>
              <TableCell>Job Title</TableCell>
              <TableCell>Stars</TableCell>
              <TableCell>Learning Progress</TableCell>
              <TableCell></TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {chefs.map(
              ({ id, jobTitle, michelinStars, name, trainingProgress }) => (
                <TableRow key={id}>
                  <TableCell>{name}</TableCell>
                  <TableCell>{jobTitle}</TableCell>
                  <TableCell>
                    <DisplayMichelinStars stars={michelinStars} />
                  </TableCell>
                  <TableCell>{trainingProgress}</TableCell>
                  <TableCell>
                    <Checkbox
                      onChange={(event) =>
                        onBoxChecked(id, event.target.checked)
                      }
                    />
                  </TableCell>
                </TableRow>
              )
            )}
          </TableBody>
        </Table>
      </StyledTableContainer>
      <Box>
        <StyledButton
          onClick={() => {
            onSelected(selectedIds);
            //We need to clear the selection every time we click because those people will be removed and aren't selectable
            setSelectedIds([]);
          }}
          disabled={selectedIds.length === 0}
          variant="outlined"
        >
          {buttonText}
        </StyledButton>
      </Box>
    </>
  );
};
