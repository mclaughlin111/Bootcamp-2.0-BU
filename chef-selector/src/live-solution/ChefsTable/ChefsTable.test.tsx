import { render, screen, within } from "@testing-library/react";
import { Chef } from "../../types";
import { ChefsTable } from "./ChefsTable";
import userEvent from "@testing-library/user-event";

const testChefs: Chef[] = [
  {
    id: 4,
    name: "Bella Bryant",
    jobTitle: "Head Chef",
    michelinStars: 2,
    trainingProgress: 0.45,
  },
  {
    id: 12,
    name: "Cade Stark",
    jobTitle: "Head Chef",
    michelinStars: 2,
    trainingProgress: 1,
  },
  {
    id: 13,
    name: "Easton McConnell",
    jobTitle: "Head Chef",
    michelinStars: 0,
    trainingProgress: 0.3,
  },
  {
    id: 8,
    name: "Hadley Farrell",
    jobTitle: "Pastry Chef",
    michelinStars: 0,
    trainingProgress: 0.25,
  },
  {
    id: 2,
    name: "Houston Dixon",
    jobTitle: "Executive Chef",
    michelinStars: 0,
    trainingProgress: 0.2,
  },
];

const clickCheckbox = async (nameRegEx: RegExp) => {
  const bellaBryantRow = screen.getByRole("row", { name: nameRegEx });
  const checkBox = within(bellaBryantRow).getByRole("checkbox");
  await userEvent.click(checkBox);
};

describe("ChefsTable", () => {
  it("should support different renders with disabled Button", () => {
    const { container } = render(
      <ChefsTable
        chefs={testChefs}
        buttonText={"Test me"}
        onSelected={() => undefined}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("should allow the user to click multiple check boxes", async () => {
    const chefsSelectMock = jest.fn();

    render(
      <ChefsTable
        buttonText={"Test"}
        chefs={testChefs}
        onSelected={chefsSelectMock}
      />
    );

    await clickCheckbox(/Bella Bryant/);
    await clickCheckbox(/Cade Stark/);

    const selectButton = screen.getByRole("button");
    await userEvent.click(selectButton);

    expect(chefsSelectMock).toHaveBeenCalledTimes(1);
    expect(chefsSelectMock).toHaveBeenCalledWith([4, 12]);
  });

  it("should allow the user to click snd unclick check boxes", async () => {
    const chefsSelectMock = jest.fn();

    render(
      <ChefsTable
        buttonText={"Test"}
        chefs={testChefs}
        onSelected={chefsSelectMock}
      />
    );

    await clickCheckbox(/Bella Bryant/);
    await clickCheckbox(/Cade Stark/);
    await clickCheckbox(/Easton McConnell/);
    await clickCheckbox(/Cade Stark/);

    const selectButton = screen.getByRole("button");
    await userEvent.click(selectButton);

    expect(chefsSelectMock).toHaveBeenCalledTimes(1);
    expect(chefsSelectMock).toHaveBeenCalledWith([4, 13]);
  });
});
