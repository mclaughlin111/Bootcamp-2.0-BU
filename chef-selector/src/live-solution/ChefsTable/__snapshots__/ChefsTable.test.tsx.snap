// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChefsTable should support different renders with disabled Button 1`] = `
<div>
  <div
    class="MuiTableContainer-root sc-gJhJTp kkwGgR css-rorn0c-MuiTableContainer-root"
  >
    <table
      class="MuiTable-root css-quj9j6-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Job Title
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Stars
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Learning Progress
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          />
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Bella Bryant
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Head Chef
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <img
              alt="michelin star"
              src="/michelin-star.png"
              width="24"
            />
            <img
              alt="michelin star"
              src="/michelin-star.png"
              width="24"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.45
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Cade Stark
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Head Chef
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <img
              alt="michelin star"
              src="/michelin-star.png"
              width="24"
            />
            <img
              alt="michelin star"
              src="/michelin-star.png"
              width="24"
            />
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            1
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Easton McConnell
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Head Chef
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          />
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.3
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Hadley Farrell
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Pastry Chef
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          />
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.25
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Houston Dixon
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Executive Chef
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          />
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            0.2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            <span
              class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            >
              <input
                class="PrivateSwitchBase-input css-1m9pwf3"
                data-indeterminate="false"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CheckBoxOutlineBlankIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="MuiBox-root css-0"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary sc-braxZu cRpLRS css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      Test me
    </button>
  </div>
</div>
`;
